### PR TITLE
sets, dicts, and sorted arrays in IR

### PR DESCRIPTION
**Pull request by catoverdrive**
adds ArraySort, Set, Dict nodes to IR. 

Sets and dicts are stored as sorted arrays in the region; Sets keep all distinct values and dicts keep all distinct keys (with non-guaranteed value if multiple items have the same key) where the (key, value) tuple isn't missing.

Map, Filter, FlatMap, Fold still need to be expanded for sets and dicts.

Builds on #3388.

### CODE REVIEW

**Comment by danking** [comment]
sorry :( I'll look Monday afternoon, I promise.

File: src/main/scala/is/hail/annotations/CodeOrdering.scala [hunk 1]
```diff
@@ -25,10 +25,29 @@ case class CodeOrdering(t: Type, missingGreatest: Boolean) {
     Code.invokeStatic[java.lang.Long, Long, Long, Int]("compare", v1, v2)
 
   private[this] def floatCompare(v1: Code[Float], v2: Code[Float]): Code[Int] =
-    Code.invokeStatic[java.lang.Float, Float, Float, Int]("compare", v1, v2)
+    v1.ceq(v2).mux(0, Code.invokeStatic[java.lang.Float, Float, Float, Int]("compare", v1, v2))
```
Comment (danking) [inline src/main/scala/is/hail/annotations/CodeOrdering.scala]: Does `Float.compare` not handle equality correctly?

File: src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala [hunk 2]
```diff
@@ -114,7 +114,8 @@ class StagedRegionValueBuilder private(val mb: MethodBuilder, val typ: Type, var
       boff := region.appendInt(bytes.length()),
       toUnit(region.appendBytes(bytes)),
       typ.fundamentalType match {
-        case _: TBinary => _empty
+        case _: TBinary =>
+          startOffset := boff
```
Comment (danking) [inline src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala]: woah, I guess this is a long-standing bug in SRVB?

Comment (danking) [reply src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala]: Do we have a matching test for this behavior?

**Comment by catoverdrive** [comment]
closing this to fix CodeOrdering behavior first

**Pull request closed without merging**